docs/DafnyRef/Types.md

```diff
@@ -2823,6 +2997,28 @@ whereas it would have been incorrect to have written something like:
 var f': (C, int, bool) -> real := F;  // not correct
 ```
 
+The arrow types themselves do not divide its parameters into ghost
+versus non-ghost. Instead, a function used as a first-class value is
+considered to be ghost if either the function or any of its arguments
+is ghost. The following example program illustrates:
```

Comment:
this is weird. I would have assumed that ghost would extend to types as well, so that there is a "compiled" version of each type.
For example, the function f below would have type

(int, ghost int) -> int

which
1) should not be assignable to a compiled (int, int) -> int
2) might be assignable to a ghost (int, ghost int) -> int
2) should be assignable to a compiled (int) -> int

Since this is the documentation of the existing implementation, there is nothing to change now, but we might want to continue the discussion. It seems that the use of ghost variables in parameters or return is used to prove complexity guarantee or proof of side-effects, so it seems legitimate to think that they will be erased so that developers would not think "oh no, I added this ghost parameter to this function, this breaks my assignment, let me create an auxiliary function then".

Reply:
Having a possible `ghost` indicator for each parameter is a sensible feature to have in the language. Using the syntax of adding the `ghost` keyword in front of each domain type, as you have shown, is nicely consistent with the syntax for ghost tuples. As you point out, designing, implementing, and documenting such a feature would be a separate Issue and PR.

Adding to the discussion of such a feature, I disagree that `f: (int, ghost int) -> int` should be assignable to `g: (int) -> int`. Such a conversion is nontrivial, since it cannot always satisfy

``` dafny
forall x: int, y: int :: f(x, y) == g(x)
```

The reason is that `f(x, y)` and `f(x, y+1)` might return different values. So, the best one could hope for is

``` dafny
forall x: int :: exists y: int :: g(x) == f(x, y)
```

and this seems too subtle to build into the language. So, I suggest that a language feature built around arrow types with ghost parameters _not_ allow automatic conversions between different arrow types.

Reply:
I did not realize until recently that methods did not have types, only functions and function methods do.